File: MtxVec.CPP/Intro.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Intro.h"
#include <Math387.hpp>
#include <MtxVec.hpp>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TIntroduction *Introduction;
//---------------------------------------------------------------------------
__fastcall TIntroduction::TIntroduction(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TIntroduction::FormCreate(TObject *Sender)
{
  Label1->Caption = "Welcome to MtxVec v" + FormatSample("0.00", MtxVecVersion / 100.0);

  RichEdit1->Lines->Clear();
  RichEdit1->SelAttributes->Style = RichEdit1->SelAttributes->Style << TFontStyle::fsBold;
  RichEdit1->SelAttributes->Color = (TColor) RGB(51, 104, 196);
  RichEdit1->SelAttributes->Size = 11;
  RichEdit1->Lines->Add("Welcome to MtxVec, an object oriented threaded and vectorized numerical library "
                        "for Delphi and C++Builder");
  RichEdit1->Lines->Add("");
  RichEdit1->Lines->Add("MtxVec adds the following capabilities to your development environment:");
  RichEdit1->Lines->Add("");
  RichEdit1->Paragraph->Numbering = nsBullet;
  RichEdit1->Paragraph->FirstIndent = 10;
  RichEdit1->Lines->Add("Comprehensive set of mathematical and statistical functions");
  RichEdit1->Lines->Add("Substantial performance improvements of floating point math by exploiting the SSE3, SSE4.x, AVX, AVX2 and AVX512 instruction sets.");
  RichEdit1->Lines->Add("Support for 64bit applications.");
  RichEdit1->Lines->Add("Certified support for 4GB address range on 64bit OS for 32bit applications.");
  RichEdit1->Lines->Add("Open CL support allows high performance cross platform portable code and gives you access to the power of GPUs.");
  RichEdit1->Lines->Add("Improved compactness and readability of code.");
  RichEdit1->Lines->Add("Significantly shorter development times by protecting the developer from a wide range of possible errors.");
  RichEdit1->Lines->Add("Same API interface in Delphi and C++Builder for W32 and W64");
  RichEdit1->Lines->Add("Support for 64bit in C++Builder");
  RichEdit1->Paragraph->Numbering = nsNone;
  RichEdit1->Paragraph->FirstIndent = 0;
  RichEdit1->Lines->Add("");
  RichEdit1->Lines->Add("Typical performance improvements observed by most users are 4-6 times for "
                        "vector functions, but speed ups up to 20 times are not rare. The matrix "
                        "multiplication for example is faster up to 50 times.");
  RichEdit1->Lines->Add("");
  RichEdit1->Lines->Add("Navigate through the demo application, to learn "
                        "more about MtxVec and experience it's unmatched "
                        "performance and ease of use.");
  RichEdit1->Lines->Add("");

  // System info section
  RichEdit1->SelAttributes->Style = RichEdit1->SelAttributes->Style << TFontStyle::fsBold;
  RichEdit1->SelAttributes->Color = (TColor) RGB(51, 104, 196);
  RichEdit1->SelAttributes->Size = 11;
  RichEdit1->Lines->Add("System info");
  RichEdit1->Lines->Add("");
  RichEdit1->Paragraph->Numbering = nsBullet;
  RichEdit1->Paragraph->FirstIndent = 10;
  RichEdit1->Lines->Add("CPU Logical Core Count = " + IntToStr((__int64)Controller->CpuCoresLogical));
  RichEdit1->Lines->Add("CPU Physical Core Count = " + IntToStr((__int64)Controller->CpuCoresPhysical));
  RichEdit1->Lines->Add("CPU NUMA Node Count = " + IntToStr((__int64)Controller->CpuNumaNodes));
  RichEdit1->Lines->Add("CPU L1 Cache size per Core = " + IntToStr((__int64)(Controller->CpuCacheL1SizeInBytes / 1024)) + " KB");
  RichEdit1->Lines->Add("CPU L2 Cache size per Core = " + IntToStr((__int64)(Controller->CpuCacheL2SizeInBytes / 1024)) + " KB");
  if (Controller->CpuCacheL3SizeInBytes > 0)
    RichEdit1->Lines->Add("CPU L3 Cache Size = " + IntToStr((__int64)(Controller->CpuCacheL3SizeInBytes / 1024)) + " KB");
  RichEdit1->Lines->Add("CPU Vendor: " + Controller->CPUVendor);

  RichEdit1->Paragraph->Numbering = nsNone;
  RichEdit1->Paragraph->FirstIndent = 0;
  RichEdit1->Lines->Add("");

  // MtxVec config section
  RichEdit1->SelAttributes->Style = RichEdit1->SelAttributes->Style << TFontStyle::fsBold;
  RichEdit1->SelAttributes->Color = (TColor) RGB(51, 104, 196);
  RichEdit1->SelAttributes->Size = 11;
  RichEdit1->Lines->Add("MtxVec config");
  RichEdit1->Lines->Add("");
  RichEdit1->Paragraph->Numbering = nsBullet;
  RichEdit1->Paragraph->FirstIndent = 10;
  RichEdit1->Lines->Add("CPU/Thread Count = " + IntToStr((__int64)Controller->CpuCores));
  RichEdit1->Lines->Add("Thread Dimension = " + IntToStr((__int64)Controller->ThreadDimension));
  RichEdit1->Lines->Add("Blas internal threads = " + IntToStr((__int64)Controller->BlasThreadCount));
  RichEdit1->Lines->Add("Lapack internal threads = " + IntToStr((__int64)Controller->LapackThreadCount));
  RichEdit1->Lines->Add("FFT internal threads = " + IntToStr((__int64)Controller->FftThreadCount));
  RichEdit1->Lines->Add("VML internal threads = " + IntToStr((__int64)Controller->VmlThreadCount));
  // IPP/MKL/FFT/VML/Random Gen acceleration flags mirror the Pascal source's
  // {$IFDEF IPPDLL_DOUBLE} etc. conditional defines. The BCB cbproj does not
  // ship these defines by default; add them to <Defines> per build to flip
  // the bullet to "ON". When undefined the build links the same DLLs via the
  // runtime packages, so a project-side define is the right toggle, not a
  // runtime probe.
  RichEdit1->Lines->Add(AnsiString("IPP acceleration (double) = ") +
                        #ifdef IPPDLL_DOUBLE
                        "ON"
                        #else
                        "OFF"
                        #endif
                        );
  RichEdit1->Lines->Add(AnsiString("IPP acceleration (single) = ") +
                        #ifdef IPPDLL_SINGLE
                        "ON"
                        #else
                        "OFF"
                        #endif
                        );
  RichEdit1->Lines->Add(AnsiString("LAPACK acceleration (double) = ") +
                        #ifdef MKLDLL_DOUBLE
                        "ON"
                        #else
                        "OFF"
                        #endif
                        );
  RichEdit1->Lines->Add(AnsiString("LAPACK acceleration (single) = ") +
                        #ifdef MKLDLL_SINGLE
                        "ON"
                        #else
                        "OFF"
                        #endif
                        );
  RichEdit1->Lines->Add(AnsiString("FFT acceleration (single and double) = ") +
                        #ifdef FFTDLL
                        "ON"
                        #else
                        "OFF"
                        #endif
                        );
  RichEdit1->Lines->Add(AnsiString("VML acceleration (double) = ") +
                        #ifdef VMLDLL_DOUBLE
                        "ON"
                        #else
                        "OFF"
                        #endif
                        );
  RichEdit1->Lines->Add(AnsiString("VML acceleration (single) = ") +
                        #ifdef VMLDLL_SINGLE
                        "ON"
                        #else
                        "OFF"
                        #endif
                        );
  RichEdit1->Lines->Add(AnsiString("Random Gen acceleration (single and double) = ") +
                        #ifdef RNDDLL
                        "ON"
                        #else
                        "OFF"
                        #endif
                        );
  RichEdit1->Lines->Add("MtxVec Block Size (double) = " + IntToStr((__int64)Math387::MtxVecBlockSize));

	TMtxVecPoolItem *p = Controller->Pool[0];
	RichEdit1->Lines->Add(AnsiString("sVector object cache: ") +
	  IntToStr((__int64)(p->Vec ? p->Vec->CacheSize : 0)) + " objects x " +
	  IntToStr((__int64)(p->Vec ? p->Vec->CacheElements : 0)) + " elements/obj");
	RichEdit1->Lines->Add(AnsiString("sVectorInt object cache: ") +
	  IntToStr((__int64)(p->VecInt ? p->VecInt->CacheSize : 0)) + " objects x " +
	  IntToStr((__int64)(p->VecInt ? p->VecInt->CacheElements : 0)) + " elements/obj");
	RichEdit1->Lines->Add(AnsiString("sMatrix object cache: ") +
	  IntToStr((__int64)(p->Mtx ? p->Mtx->CacheSize : 0)) + " objects x " +
	  IntToStr((__int64)(p->Mtx ? p->Mtx->CacheElements : 0)) + " elements/obj");
	RichEdit1->Lines->Add(AnsiString("sMatrixInt object cache: ") +
	  IntToStr((__int64)(p->MtxInt ? p->MtxInt->CacheSize : 0)) + " objects x " +
	  IntToStr((__int64)(p->MtxInt ? p->MtxInt->CacheElements : 0)) + " elements/obj");

  RichEdit1->Paragraph->Numbering = nsNone;
  RichEdit1->Paragraph->FirstIndent = 0;

  RichEdit1->Lines->Add("");

  // Enabled CPU support
  RichEdit1->SelAttributes->Style = RichEdit1->SelAttributes->Style << TFontStyle::fsBold;
  RichEdit1->SelAttributes->Color = (TColor) RGB(51, 104, 196);
  RichEdit1->SelAttributes->Size = 11;
  RichEdit1->Lines->Add("Enabled CPU support");
  RichEdit1->Lines->Add("");
  RichEdit1->Paragraph->Numbering = nsBullet;
  RichEdit1->Paragraph->FirstIndent = 10;

  if (Controller->CpuSSSE3) RichEdit1->Lines->Add("SSSE3");
  if (Controller->CpuSSE42) RichEdit1->Lines->Add("SSE42");
  if (Controller->CpuAVX) RichEdit1->Lines->Add("AVX");
  if (Controller->CpuAVX2) RichEdit1->Lines->Add("AVX2");
  if (Controller->CpuAVX512F) RichEdit1->Lines->Add("AVX512F");
  if (Controller->CpuAVX512CD) RichEdit1->Lines->Add("AVX512CD");
  if (Controller->CpuAVX512BW) RichEdit1->Lines->Add("AVX512BW");
  if (Controller->CpuAVX512DQ) RichEdit1->Lines->Add("AVX512DQ");

  if (!Controller->CpuFeaturesEnabledMatchAvailable())
    ShowMessage("Warning: The CPU features enabled do not match the CPU features available!");

  RichEdit1->Paragraph->Numbering = nsNone;
  RichEdit1->Paragraph->FirstIndent = 0;
  RichEdit1->Lines->Add("");
}
//---------------------------------------------------------------------------

File: MtxVec.Delphi/Intro.pas
unit Intro;

{$I bdsppdefs.inc}

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, ComCtrls, jpeg,  ExtCtrls;

type
  TIntroduction = class(TForm)
    RichEdit1: TRichEdit;
    Label1: TLabel;
    Panel1: TPanel;
    Image1: TImage;
    Image2: TImage;
    Image3: TImage;
    procedure FormCreate(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
  end;

var
  Introduction: TIntroduction;

implementation

uses Math387, MtxVec, ippsplTypes
     {$IFDEF IPPDLL_DOUBLE}
     ,ippspl
     {$ENDIF}
     {$IFDEF IPPDLL_SINGLE}
     ,ippsplSingle
     {$ENDIF}
     ;

{$IFDEF CLR}
{$R *.NFM}
{$ELSE}
{$R *.DFM}
{$ENDIF}

procedure TIntroduction.FormCreate(Sender: TObject);
{$IFDEF IPPDLL_ANY}
var aMask, bMask: Int64;
{$ENDIF}
begin
  Label1.Caption := 'Welcome to MtxVec v' + FormatSample('0.00',MtxVecVersion/100);
  With RichEdit1.Lines, RichEdit1 do
  begin
    Clear;
    SelAttributes.Style := [fsBold];
    SelAttributes.Color := RGB(51,104,196);
    SelAttributes.Size := 11;
    Add('Welcome to MtxVec, an object oriented threaded and vectorized numerical library '
      + 'for Delphi and .NET platform');
    Add('');
    Add('MtxVec adds the following capabilities to your development environment:');
    Add('');
    Paragraph.Numbering := nsBullet;
    Paragraph.FirstIndent := 10;
    Add('Comprehensive set of mathematical and statistical functions');
    Add('Substantial performance improvements of floating point math by exploiting the SSE3, SSE4.x, AVX, AVX2 and AVX512 instruction sets.');
    Add('Support for 64bit applications.');
    Add('Certified support for 4GB address range on 64bit OS for 32bit applications.');
    Add('Open CL support allows high performance cross platform portable code and gives you access to the power of GPUs.');
    Add('Improved compactness and readability of code.');
    Add('Significantly shorter development times by protecting the developer from a wide range of possible errors.');
    Add('Same API interface in Delphi for W32 and in .NET');
    Add('Support for 64bit in .NET');
    Paragraph.Numbering := nsNone;
    Paragraph.FirstIndent := 0;
    Add('');
    Add('Typical performance improvements observed by most users are 4-6 times for ' +
        'vector functions, but speed ups up to 20 times are not rare. The matrix ' +
        'multiplication for example is faster up to 50 times.');
    Add('');
    Add('Navigate through the demo application, to learn '
      + 'more about MtxVec and experience it''s unmatched '
      + 'performance and ease of use.');
    Add('');
    SelAttributes.Style := [fsBold];
    SelAttributes.Color := RGB(51,104,196);
    SelAttributes.Size := 11;
    Add('System info');
    Add('');
    Paragraph.Numbering := nsBullet;
    Paragraph.FirstIndent := 10;
    Add('CPU Logical Core Count = ' + IntToStr(Controller.CpuCoresLogical));
    Add('CPU Physical Core Count = ' + IntToStr(Controller.CpuCoresPhysical));
    Add('CPU NUMA Node Count = ' + IntToStr(Controller.CpuNumaNodes));
    Add('CPU L1 Cache size per Core = ' + IntToStr(Controller.CpuCacheL1SizeInBytes div 1024) + ' KB');
    Add('CPU L2 Cache size per Core = ' + IntToStr(Controller.CpuCacheL2SizeInBytes div 1024) + ' KB');
    if Controller.CpuCacheL3SizeInBytes > 0 then Add('CPU L3 Cache Size = ' + IntToStr(Controller.CpuCacheL3SizeInBytes div 1024) + ' KB');
    Add('CPU Vendor: ' + Controller.CPUVendor);

    Paragraph.Numbering := nsNone;
    Paragraph.FirstIndent := 0;
    Add('');
    SelAttributes.Style := [fsBold];
    SelAttributes.Color := RGB(51,104,196);
    SelAttributes.Size := 11;
    Add('MtxVec config');
    Add('');
    Paragraph.Numbering := nsBullet;
    Paragraph.FirstIndent := 10;
    Add('CPU/Thread Count = ' + IntToStr(Controller.CPUCores));
    Add('Thread Dimension = ' + IntToStr(Controller.ThreadDimension));
    Add('Blas internal threads = ' + IntToStr(Controller.BlasThreadCount));
    Add('Lapack internal threads = ' + IntToStr(Controller.LapackThreadCount));
    Add('FFT internal threads = ' + IntToStr(Controller.FFTThreadCount));
    Add('VML internal threads = ' + IntToStr(Controller.VMLThreadCount));
    Add('IPP acceleration (double) = ' +  {$IFDEF IPPDLL_DOUBLE} 'ON' {$ELSE} 'OFF' {$ENDIF});
    Add('IPP acceleration (single) = ' +  {$IFDEF IPPDLL_SINGLE} 'ON' {$ELSE} 'OFF' {$ENDIF});
    Add('LAPACK acceleration (double) = ' +  {$IFDEF MKLDLL_DOUBLE} 'ON' {$ELSE} 'OFF' {$ENDIF});
    Add('LAPACK acceleration (single) = ' +  {$IFDEF MKLDLL_SINGLE} 'ON' {$ELSE} 'OFF' {$ENDIF});
    Add('FFT acceleration (single and double) = ' +  {$IFDEF FFTDLL} 'ON' {$ELSE} 'OFF' {$ENDIF});
    Add('VML acceleration (double) = ' +  {$IFDEF VMLDLL_DOUBLE} 'ON' {$ELSE} 'OFF' {$ENDIF});
    Add('VML acceleration (single) = ' +  {$IFDEF VMLDLL_SINGLE} 'ON' {$ELSE} 'OFF' {$ENDIF});
    Add('Random Gen acceleration (single and double) = ' +  {$IFDEF RNDDLL} 'ON' {$ELSE} 'OFF' {$ENDIF});
    Add('MtxVec Block Size (double) = ' + IntToStr(Math387.MtxVecBlockSize));
    Paragraph.Numbering := nsNone;
    Paragraph.FirstIndent := 0;

    {$IFDEF IPPDLL_ANY}
    Add('');
    SelAttributes.Style := [fsBold];
    SelAttributes.Color := RGB(51,104,196);
    SelAttributes.Size := 11;
    Add('Enabled CPU support');
    Add('');
    Paragraph.Numbering := nsBullet;
    Paragraph.FirstIndent := 10;

    aMask := ippGetEnabledCpuFeatures;
    if (aMask and ippsplTypes.ippCPUID_MMX) <> 0 then Add('MMX');
    if (aMask and ippsplTypes.ippCPUID_SSE) <> 0 then Add('SSE');
    if (aMask and ippsplTypes.ippCPUID_SSE2) <> 0 then Add('SSE2');
    if (aMask and ippsplTypes.ippCPUID_SSSE3) <> 0 then Add('SSSE3');
    if (aMask and ippsplTypes.ippCPUID_SSE42) <> 0 then Add('SSE42');
    if (aMask and ippsplTypes.ippCPUID_AVX) <> 0 then Add('AVX');
    if (aMask and ippsplTypes.ippCPUID_AVX2) <> 0 then Add('AVX2');
    if (aMask and ippsplTypes.ippCPUID_AVX512F) <> 0 then Add('AVX512F');
    if (aMask and ippsplTypes.ippCPUID_AVX512CD) <> 0 then Add('AVX512CD');
    if (aMask and ippsplTypes.ippCPUID_AVX512BW) <> 0 then Add('AVX512BW');
    if (aMask and ippsplTypes.ippCPUID_AVX512DQ) <> 0 then Add('AVX512DQ');
//    if (aMask and ippsplTypes.ippCPUID_AVX512PF) <> 0 then Add('AVX512PF');
//    if (aMask and ippsplTypes.ippCPUID_AVX512ER) <> 0 then Add('AVX512ER');
//    if (aMask and ippsplTypes.ippCPUID_AVX512VL) <> 0 then Add('AVX512VL');

    ippGetCpuFeatures(bMask, nil);

    if bMask <> aMask then ShowMessage('Warning: The CPU features enabled do not match the CPU features available!');

    {$ENDIF}

    Paragraph.Numbering := nsNone;
    Paragraph.FirstIndent := 0;
    Add('');
  end;
end;

initialization
  RegisterClass(TIntroduction);
end.
